File: LastApple.Web/ClientApp-Svelte/src/lib/api/lastfmApi.ts
// Last.fm API client
import environment from '$lib/services/environment';

const API_URL = environment.apiUrl;

export interface ILastfmUser {
    readonly name: string;
    readonly url: string;
    readonly avatar: Array<string>;
}

interface ILastfmArtist {
    name: string;
    url?: string;
}

function getSessionId(): string | null {
    if (typeof localStorage === 'undefined') return null;
    return localStorage.getItem('SessionId');
}

const lastfmApi = {
    async getAuthUrl(redirectUrl: string): Promise<string> {
        const encodedUrl = encodeURIComponent(redirectUrl);
        const response = await fetch(`${API_URL}api/lastfm/auth?redirectUrl=${encodedUrl}`);
        return await response.json();
    },

    async searchArtist(searchTerm: string): Promise<ILastfmArtist[]> {
        const response = await fetch(`${API_URL}api/lastfm/artist/search?term=${searchTerm}`);
        return await response.json();
    },

    async searchTag(searchTerm: string): Promise<{ name: string }[]> {
        const response = await fetch(`${API_URL}api/lastfm/tag/search?term=${searchTerm}`);
        return await response.json();
    },

    async postToken(token: string): Promise<string> {
        const response = await fetch(`${API_URL}api/lastfm/auth?token=${token}`, {
            method: 'POST',
            headers: { 'X-SessionId': getSessionId() || '' }
        });
        return await response.json();
    },

    async logout(): Promise<Response> {
        return await fetch(`${API_URL}api/lastfm/auth`, {
            method: 'DELETE',
            headers: { 'X-SessionId': getSessionId() || '' }
        });
    },

    async getUser(): Promise<ILastfmUser> {
        const response = await fetch(`${API_URL}api/lastfm/auth/user`, {
            headers: { 'X-SessionId': getSessionId() || '' }
        });
        return await response.json();
    },

    async scrobble(song: string, artist: string, album?: string, timestamp?: number): Promise<void> {
        await fetch(`${API_URL}api/lastfm/scrobble`, {
            method: 'POST',
            headers: {
                'X-SessionId': getSessionId() || '',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ artist, song, album, timestamp })
        });
    },

    async updateNowPlaying(song: string, artist: string, album?: string): Promise<void> {
        await fetch(`${API_URL}api/lastfm/nowPlaying`, {
            method: 'POST',
            headers: {
                'X-SessionId': getSessionId() || '',
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ artist, song, album })
        });
    }
};

export default lastfmApi;

// Named exports for convenience
export const { getAuthUrl, searchArtist, searchTag, postToken, logout, getUser, scrobble, updateNowPlaying } = lastfmApi;

File: Lastream.iOS/Lastream/Services/API/LastfmAPI.swift
import Foundation

/// Protocol for Last.fm API operations.
protocol LastfmAPIProtocol: Sendable {
    /// Gets the Last.fm OAuth authorization URL.
    func getAuthUrl(redirectUrl: String) async throws -> URL
    
    /// Completes the OAuth flow by posting the authorization token.
    func postToken(_ token: String) async throws -> String
    
    /// Logs out from Last.fm.
    func logout() async throws
    
    /// Gets the authenticated Last.fm user.
    func getUser() async throws -> LastfmUser?
    
    /// Searches for artists by name.
    func searchArtist(term: String) async throws -> [LastfmArtist]
    
    /// Searches for tags by name.
    func searchTag(term: String) async throws -> [LastfmTag]
    
    /// Scrobbles a track to Last.fm.
    func scrobble(song: String, artist: String, album: String?, timestamp: Date?) async throws
    
    /// Updates the "now playing" status on Last.fm.
    func updateNowPlaying(song: String, artist: String, album: String?) async throws
}

/// Request body for scrobble and now playing endpoints.
struct ScrobbleRequest: Codable, Sendable {
    let artist: String
    let song: String
    let album: String?
    let timestamp: Int?
    
    init(artist: String, song: String, album: String? = nil, timestamp: Date? = nil) {
        self.artist = artist
        self.song = song
        self.album = album
        self.timestamp = timestamp.map { Int($0.timeIntervalSince1970) }
    }
}

/// Implementation of the Last.fm API.
final class LastfmAPI: LastfmAPIProtocol, Sendable {
    private let client: APIClientProtocol
    
    init(client: APIClientProtocol) {
        self.client = client
    }
    
    func getAuthUrl(redirectUrl: String) async throws -> URL {
        let encodedUrl = redirectUrl.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? redirectUrl
        let urlString: String = try await client.getString("api/lastfm/auth?redirectUrl=\(encodedUrl)")
        
        guard let url = URL(string: urlString) else {
            throw APIError.invalidResponse
        }
        return url
    }
    
    func postToken(_ token: String) async throws -> String {
        try await client.post("api/lastfm/auth?token=\(token)")
    }
    
    func logout() async throws {
        try await client.delete("api/lastfm/auth")
    }
    
    func getUser() async throws -> LastfmUser? {
        do {
            return try await client.get("api/lastfm/auth/user")
        } catch APIError.notFound {
            return nil
        } catch {
            // If we get an empty response or null, return nil
            return nil
        }
    }
    
    func searchArtist(term: String) async throws -> [LastfmArtist] {
        let encodedTerm = term.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? term
        return try await client.get("api/lastfm/artist/search?term=\(encodedTerm)")
    }
    
    func searchTag(term: String) async throws -> [LastfmTag] {
        let encodedTerm = term.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? term
        return try await client.get("api/lastfm/tag/search?term=\(encodedTerm)")
    }
    
    func scrobble(song: String, artist: String, album: String?, timestamp: Date?) async throws {
        let request = ScrobbleRequest(artist: artist, song: song, album: album, timestamp: timestamp)
        try await client.post("api/lastfm/scrobble", body: request) as EmptyResponse
    }
    
    func updateNowPlaying(song: String, artist: String, album: String?) async throws {
        let request = ScrobbleRequest(artist: artist, song: song, album: album)
        try await client.post("api/lastfm/nowplaying", body: request) as EmptyResponse
    }
}

/// Empty response type for endpoints that don't return data.
private struct EmptyResponse: Decodable {}
